### PR TITLE
honor `server.ProxyProtocol` in destination service

### LINKED ISSUE #6857: Honor `server.proxyProtocol=opaque` in client discovery
Server resources have a `proxyProtocol` field that can be set to _opaque_. This setting can eliminate the need to configure the annotation `config.linkerd.io/opaque-ports`. We should update the destination controller to use a pod:port's server configuration when possible.

Naively, we might have to update the destination controller to replicate the indexing logic in the policy controller; but, instead, we can probably have the destination controller establish a watch (via grpc over localhost for each pod and port). Note that we really only need to establish a watch for each pod-port that is actually resolved--most pods are only resolved through their `service` IP, where the destination controller uses the _service's_ opaque annotation. `Server` resources are only used when a client resolves a pod IP directly. Alternatively, we could add new bulk API endpoints to the controller API; but this may be overkill for now.

### PR DESCRIPTION
**Pull request by kleimkuhler**
Currently, the destination service only looks at the `config.linkerd.io/opaque-ports` annotation on a pod; if `port` is in the annotation value, then the destination profile indicates it supports opaque traffic . Now that `Server` resources can _also_ indicate a destination supports opaque traffic, the destination service must consider these as well.

Closes #6857.

When `GetProfile` is called for a pod, it has not subscribed the stream for updates. However, it now creates a policy server client which will send updates as a workload's `Server` resource changes. This takes place in `sendEndpointProfile`.

An initial destination profile is still sent which considers only the opaque ports annotation value. After that, if the destination is a pod, it will continuously check for `Server` updates or handle server or stream shutdown.

### testing
Install Linkerd, emojivoto, and emojivoto policy.

Then run the policy controller and destination server.
```
$ cargo run --no-default-features --features=native-tls
```
```
$ go run controller/cmd/main.go destination -kubeconfig ~/.kube/config
```

Now get the destination profile for the `web` pod.
```
$ go run controller/script/destination-client/main.go -method getProfile -path 10.42.0.23:8080
...
INFO[0000] retry_budget:{retry_ratio:0.2 min_retries_per_second:10 ttl:{seconds:10}} endpoint:{addr:{ip:{ipv4:170524695} port:8080} weight:10000 metric_labels:{key:"control_plane_ns" value:"linkerd"} metric_labels:{key:"deployment" value:"web"} metric_labels:{key:"namespace" value:"emojivoto"} metric_labels:{key:"pod" value:"web-5f86686c4d-qlhdk"} metric_labels:{key:"pod_template_hash" value:"5f86686c4d"} metric_labels:{key:"serviceaccount" value:"web"} tls_identity:{dns_like_identity:{name:"web.emojivoto.serviceaccount.identity.linkerd.cluster.local"}} protocol_hint:{h2:{}}}
```

Observe that the destination profile does not have `opaque_protocol=true` and the endpoint does not hint it supports opaque traffic.

Now, update the `web-http` server so that it's protocol is opaque.
```
kubectl edit -n emojivoto server web-http
...
  proxyProtocol: opaque
```

Observe that the new destination profile has `opaque_protocol=true` and the hint properly set.
```
INFO[0211] opaque_protocol:true ... endpoint:{... protocol_hint:{h2:{} opaque_transport:{inbound_port:8080}}}
```

If you edit `web-http` back to `HTTP/1`, a new destination profile should be received that mirrors the first.

Signed-off-by: Kevin Leimkuhler <kevin@kleimkuhler.com>

### CODE REVIEW

File: controller/cmd/destination/main.go [hunk 1]
```diff
@@ -24,6 +24,7 @@ func Main(args []string) {
 
 	addr := cmd.String("addr", ":8086", "address to serve on")
 	metricsAddr := cmd.String("metrics-addr", ":9996", "address to serve scrapable metrics on")
+	policyAddr := cmd.String("policy-addr", ":8090", "address to connect to the policy server")
```
Comment (olix0r) [inline controller/cmd/destination/main.go:27]:
```suggestion
	policyAddr := cmd.String("policy-addr", "127.0.0.1:8090", "address to connect to the policy server")
```

I don't think we can use `:8090` as a client's target address.

File: controller/api/destination/server_test.go [hunk 2]
```diff
@@ -917,8 +944,8 @@ func TestGetProfiles(t *testing.T) {
 		if first.Endpoint.ProtocolHint == nil {
 			t.Fatalf("Expected protocol hint but found none")
 		}
-		if first.Endpoint.ProtocolHint.GetOpaqueTransport().InboundPort != 4143 {
-			t.Fatalf("Expected pod to support opaque traffic on port 4143")
+		if first.Endpoint.ProtocolHint.GetOpaqueTransport().InboundPort != opaquePort {
+			t.Fatalf("Expected pod to support opaque traffic on port %d", opaquePort)
```
Comment (kleimkuhler) [inline controller/api/destination/server_test.go:948]: `InboundPort` should be the application port, not the proxy port. I'm not sure how we weren't catching this before.

File: controller/api/destination/server.go [hunk 3]
```diff
@@ -378,23 +383,94 @@ func (s *server) sendEndpointProfile(stream pb.Destination_GetProfileServer, pod
...
 		sp.Spec.OpaquePorts = opaquePorts
 		translator.Update(&sp)
 	} else {
 		translator.Update(nil)
 	}
 
+	if portSpec != nil {
+		// Create a client which watches for changes to the inbound policy for
+		// portSpec.
+		client, err := s.policyClient.WatchPort(stream.Context(), portSpec)
```
Comment (kleimkuhler) [inline controller/api/destination/server.go:415]: By using `stream.Context()`, when the stream shuts down we will encounter in an error in `client.Recv()` which leads to that Go routine gracefully shutting down.

File: controller/api/destination/endpoint_translator.go [hunk 4]
```diff
@@ -428,5 +428,5 @@ func getInboundPort(podSpec *corev1.PodSpec) (uint32, error) {
 			return uint32(port), nil
 		}
 	}
-	return 0, fmt.Errorf("failed to find %s environment variable in any container for given pod spec", envInboundListenAddr)
```
Comment (olix0r) [inline controller/api/destination/endpoint_translator.go:431]: Is this change necessary?

Comment (kleimkuhler) [reply controller/api/destination/endpoint_translator.go:431]: Yep we don't want an error returned [here](https://github.com/linkerd/linkerd2/pull/6889/files#diff-6598a7b83bed9beb2b591cedbac5b5c052a813553d4e38437494128168d7a438R391). This means the destination pod is not injected, so we don't have a proxy port to set on the protocol hint [here](https://github.com/linkerd/linkerd2/pull/6889/files#diff-281e603e6940b45931c7339138278cec5c31c347e4d3a8d336b72d9c38d0f4a1R58).

Comment (kleimkuhler) [reply controller/api/destination/endpoint_translator.go:431]: I considered making this more explicit by having `toWeightedAddr` return an additional `bool` that says if the pod is injected or not—something like `addr, injected, err := toWeightedAddr(...)`. Then when we create the `Endpoint`, we'd add a third field to it:
```
type Endpoint struct {
	addr      *pb.WeightedAddr
	proxyPort uint32
        injected bool
}
```

Finally, we'd use `endpoint.injected` instead of `endpoint.proxyPort != 0` as the check.

I decided against that though because I think it's enough to assume that if `proxyPort == 0`, the endpoint is not injected.

Comment (olix0r) [reply controller/api/destination/endpoint_translator.go:431]: where do we test for proxyPort == 0?

Comment (kleimkuhler) [reply controller/api/destination/endpoint_translator.go:431]: I can update this to clarify the return, but `toWeightedAddr` hasn't changed as part of this.

**Comment by adleong** [comment]
Some high level questions to start: up until now, it was my assumption that Server resources ONLY configured server (i.e. inbound) policy.  It shakes up my mental model a bit to see Server resources affecting client (i.e. outbound) policy.

I notice in the linked issue that it's mentioned that this only affects clients which are addressing a pod directly and not clients which are addressing a service cluster ip.  Why is that the case?  For example, when an outbound proxy gets a connection for a clusterIP service, it first does a profile lookup (which will get a profile response without an endpoint) and then an endpoints lookup.  Each endpoint returned from the endpoints lookup may or may not have the requested port defined as opaque and thus may or may not have the opaque_transport protocol hint.  If that endpoint is a pod with a Server defined for the requested port, and that Server describes the protocol as opaque, should the destination controller set the opaque_transport protocol hint on that endpoint?

File: controller/api/destination/server.go [hunk 5]
```diff
@@ -371,30 +376,114 @@ func (s *server) sendEndpointProfile(stream pb.Destination_GetProfileServer, pod
...
+			case <-s.shutdown:
+				return nil
+			case <-stream.Context().Done():
+				log.Debugf("GetProfile %s:%s:%d cancelled", pod.Namespace, pod.Name, port)
+				return nil
+			case update := <-updates:
+				if update.GetProtocol().GetOpaque() != nil {
+					if _, ok := opaquePorts[port]; !ok {
+						opaquePorts[port] = struct{}{}
+						overridden = true
```
Comment (adleong) [inline controller/api/destination/server.go:468]: I find it very confusing to maintain this state keeping track of whether this has been overridden and also plumbing similar statekeeping through the ProfileTranslator.

What if, instead, we simply update the translator's endpoint?  i.e. when we get a policy update we use that to build a new opaquePorts set and a new WeightedAddr:

```
opaquePortsWithServerProtocol := make(map[uint32]struct{})
for k, v := range opaquePorts {
  opaquePortsWithServerProtocol[k] = v
}
if update.GetProtocol().GetOpaque() != nil {
  opaquePortsWithServerProtocol[port] = struct{}{}
}
newEndpoint, err = toWeightedAddr(podSet.Addresses[podID], opaquePortsWithServerProtocol, skippedInboundPorts, s.enableH2Upgrade, s.identityTrustDomain, s.controllerNS, s.log)
```

then we can update the translator's endpoint in place

```
*endpoint = *newEndpoint
```

and then send an update

```
if len(opaquePortsWithServerProtocol) != 0 {
// Send a destination profile update that includes the opaque
// ports.
		sp := sp.ServiceProfile{}
		sp.Spec.OpaquePorts = opaquePortsWithServerProtocol
		translator.Update(&sp)
	} else {
		translator.Update(nil)
	}
```

Doing it this way should avoid us needing to make any changes to the profile translator code at all.

Comment (kleimkuhler) [reply controller/api/destination/server.go]: Great suggestion this worked well. I was able to remove all the changes from the profile translator.

File: controller/api/destination/endpoint_translator.go [hunk 6]
```diff
@@ -287,6 +303,120 @@ func (et *endpointTranslator) sendClientAdd(set watcher.AddressSet) {
...
+		}
+		et.log.Infof("Establishing watch on policy server %s:%d", policyPortSpec.Workload, policyPortSpec.Port)
+
+		// policyPb.PortSpec cannot be used as a map key, so we create an
+		// equivalent port spec with comparable keys.
+		portSpec := &portSpec{
+			workload: policyPortSpec.Workload,
+			port:     policyPortSpec.Port,
+		}
+		et.policyWatches[*portSpec] = client
```
Comment (adleong) [inline controller/api/destination/endpoint_translator.go:357]: do we need to worry about if there is already a policy watch in the map for this portSpec?  as written, I think the previous watch would be overwritten in the map and thus be leaked and no longer be cancellable.  If we add checking to see if an entry already exists in the map, we probably also have to add some locking to avoid race conditions.

File: controller/api/destination/endpoint_translator.go [hunk 7]
```diff
@@ -287,6 +303,120 @@ func (et *endpointTranslator) sendClientAdd(set watcher.AddressSet) {
...
+		go func() {
+			for {
+				select {
+				case <-done:
+					et.log.Infof("Stopping policy server watch for %s:%d", portSpec.workload, portSpec.port)
+					return
+				case <-et.stream.Context().Done():
+					et.log.Infof("Stopping policy server watch for %s:%d", portSpec.workload, portSpec.port)
+					return
+				case update := <-updates:
```
Comment (adleong) [inline controller/api/destination/endpoint_translator.go:392]: instead of having a separate done channel, can you use the 2 value receive here to check if the updates channel is close? ` case update, ok := <- updates:`  and then close the updates channel when client.Recv encounters an error.

File: controller/api/destination/endpoint_translator.go [hunk 8]
```diff
@@ -287,6 +303,120 @@ func (et *endpointTranslator) sendClientAdd(set watcher.AddressSet) {
...
+		// When an update is available, create a new weighted addr that may
+		// include an opaque transport port as specified by a policy server.
+		addr := addr
+		go func() {
+			for {
+				select {
+				case <-done:
+					et.log.Infof("Stopping policy server watch for %s:%d", portSpec.workload, portSpec.port)
+					return
+				case <-et.stream.Context().Done():
```
Comment (adleong) [inline controller/api/destination/endpoint_translator.go:389]: is this case redundant?  the context passed to the policy client is derived from the stream context which means that when the stream context becomes done, the policy request should be cancelled and client.Recv should return an error.

File: controller/api/destination/endpoint_translator.go [hunk 9]
```diff
@@ -102,9 +116,11 @@ func (et *endpointTranslator) sendFilteredUpdate(set watcher.AddressSet) {
 
 	if len(diffAdd.Addresses) > 0 {
 		et.sendClientAdd(diffAdd)
+		et.watchEndpointPolicy(diffAdd)
```
Comment (adleong) [inline controller/api/destination/endpoint_translator.go:119]: When we subscribe, we'll hit this with a set of addresses and send one big Add message containing all the endpoints in the service.  Then we'll subscribe to the policy for each of the endpoints and send an additional individual Add for each endpoint.  (and then continue to send an individual Add every time policy changes).

I'm not sure exactly how we'd want this to work, but this seems redundant and inefficient.  Perhaps we need some kind of buffering strategy to group the individual Adds into a single message?  Or maybe we can keep them as-is but instead forgo the initial Add of the entire address set?

Comment (adleong) [inline controller/api/destination/endpoint_translator.go:119]: we probably need locking around watchEndpointPolicy and  closeEndpointPolicy to avoid race conditions

**Review by adleong — COMMENTED** [review]
The `sendEndpointProfile` side of this looks correct to me.  Some questions about the endpoints translator side of things.

File: controller/api/destination/endpoint_translator.go [hunk 10]
```diff
@@ -310,6 +440,30 @@ func (et *endpointTranslator) sendClientRemove(set watcher.AddressSet) {
...
+		// server watch to close.
+		if addr.Pod == nil {
+			continue
+		}
+		portSpec := portSpec{
+			workload: fmt.Sprintf("%s:%s", addr.Pod.Namespace, addr.Pod.Name),
+			port:     addr.Port,
+		}
+		if client, ok := et.policyWatches[portSpec]; ok {
+			err := client.CloseSend()
```
[the remark this answers is not in the document]
Comment (kleimkuhler) [reply controller/api/destination/endpoint_translator.go:458]: Reading a little more about this function online it doesn't seem like it's the right thing to use here even though I was observing `Recv` errors on the other end; it looks like it still leaves some resources behind. Instead of mapping to policy server clients, I created a `watchPortClient` wrapper that includes the `cancel` function for the context and we use that now.

File: controller/api/destination/endpoint_translator.go (hunk 8 above)
Comment (kleimkuhler) [reply controller/api/destination/endpoint_translator.go:389]: Yep this is now redundant especially with handling channel closure with `update, ok := <- updates` in the other remaining case.

File: controller/api/destination/endpoint_translator.go (hunk 6 above)
Comment (kleimkuhler) [reply controller/api/destination/endpoint_translator.go]: I wrapped the policy watches with a new `policyWatches` struct that holds the watches and a mutex. `watchEndpointPolicy` and `closeEndpointPolicy` now both lock `policyWatches` before adding/deleting clients from the `watches` map.

File: controller/api/destination/endpoint_translator.go (hunk 9 above)
Comment (kleimkuhler) [reply controller/api/destination/endpoint_translator.go:135]: refer to this [comment](https://github.com/linkerd/linkerd2/pull/6889#discussion_r712674796)

Comment (kleimkuhler) [reply controller/api/destination/endpoint_translator.go:135]: I've removed this "multiphase" updating in recent commits. `sendClientAdd` now only sends add updates when dealing with addresses not backed by pods. `watchEndpointPolicy` only sends add updates for addresses backed by pods. This does mean that that adds for pods now occur individually instead of grouped together, but the first destination profile returned does now take into account server resources.

File: controller/api/destination/endpoint_translator.go [hunk 11]
```diff
@@ -211,69 +243,55 @@ func (et *endpointTranslator) NoEndpoints(exists bool) {
...
-			et.log.Errorf("Failed to translate endpoints to weighted addr: %s", err)
-			continue
-		}
 		addrs = append(addrs, wa)
 	}
 
+	// If there are no addresses to add then don't send an update.
+	if len(addrs) == 0 {
+		return
+	}
```
Comment (kleimkuhler) [inline controller/api/destination/endpoint_translator.go:293]: We don't want to send an add update with no addresses because the `add` created below does still have labels set. This results in the first add containing only labels and no addresses.

File: controller/api/destination/server.go [hunk 12]
```diff
@@ -344,56 +349,107 @@ func (s *server) GetProfile(dest *pb.GetDestination, stream pb.Destination_GetPr
...
-		var ok bool
-		opaquePorts, ok, err = getPodOpaquePortsAnnotations(pod)
-		if err != nil {
-			log.Errorf("failed to get opaque ports annotation for pod: %s", err)
-		}
+	if pod == nil {
+		translator := newProfileTranslator(stream, s.log, "", port, nil)
+		translator.Update(nil)
+		return nil
+	}
```
Comment (kleimkuhler) [inline controller/api/destination/server.go:356]: This isn't new logic. It just allows `sendEndpointProfile` to return early and reduce the nesting below.

File: controller/script/destination-client/main.go [hunk 13]
```diff
@@ -68,6 +68,9 @@ func get(client pb.DestinationClient, req *pb.GetDestination) {
 				default:
 					log.Printf("  - protocol hint: UNKNOWN")
 				}
+				if ot := addr.GetProtocolHint().GetOpaqueTransport(); ot != nil {
+					log.Printf("  - opaque transport port: %d", ot.GetInboundPort())
+				}
```
Comment (kleimkuhler) [inline controller/script/destination-client/main.go:73]: This change is related to running the destination service locally and checking the `opaque_transport` field for destination profile updates.

**Review by kleimkuhler — COMMENTED** [review]
A lot of these changes are related to change the destination tests to use channels instead of slices. Now that updates can occur in goroutines, there were race conditions with a goroutine trying to add to a slice and a test trying to read from it.

File: controller/api/destination/endpoint_translator.go [hunk 14]
```diff
@@ -287,6 +307,154 @@ func (et *endpointTranslator) sendClientAdd(set watcher.AddressSet) {
...
+				select {
+				case <-et.stream.Context().Done():
+					// This case handles shutdown when updates will never be
+					// closed. This occurs when the policy server is unable to
+					// find a server for a pod. After one sending one update,
+					// it stops checking for updates, so it will never try to
+					// close updates.
+					et.log.Debugf("Stopping policy server watch for %s:%d", portSpec.workload, portSpec.port)
+					return
+				case update, ok := <-updates:
```
Comment (adleong) [inline controller/api/destination/endpoint_translator.go:425]: I we get rid of the previous case as redundant, only one case remains in this select.  I think this means that we can get rid of this goroutine entirely and fold it into the goroutine that blocks on Revc.  In other words, we can just have one goroutine which blocks on Revc and then does all the appropriate logic (such as computing the weighted addr and calling Send).  If the policy request is ever cancelled (because the destination request was cancelled or because we explicitly call the cancelFunc) then Recv should return with an error; thus we should never get stuck.

File: controller/api/destination/endpoint_translator.go [hunk 15]
```diff
@@ -287,6 +307,154 @@ func (et *endpointTranslator) sendClientAdd(set watcher.AddressSet) {
...
+			cancel: cancel,
+		}
+
+		// Before adding the new port client, ensure any previous client for
+		// the port spec has been closed.
+		if pc, ok := et.policyWatches.watches[*portSpec]; ok {
+			pc.cancel()
+		}
+
+		et.policyWatches.watches[*portSpec] = &portClient
```
Comment (adleong) [inline controller/api/destination/endpoint_translator.go:378]: What's the intention with this cache of watches?  Right now if two different requests come in which include the same pod+port then the later policy watch will replace the earlier one in this map and the earlier one will no longer be cancelable.

Is the goal to share the same policy watch between both requests?  If so, how do updates from the policy watch get fanned out to both destination requests?

File: controller/api/destination/endpoint_translator.go [hunk 16]
```diff
@@ -310,6 +478,33 @@ func (et *endpointTranslator) sendClientRemove(set watcher.AddressSet) {
...
+		}
+		portSpec := portSpec{
+			workload: fmt.Sprintf("%s:%s", addr.Pod.Namespace, addr.Pod.Name),
+			port:     addr.Port,
+		}
+		if portClient, ok := et.policyWatches.watches[portSpec]; ok {
+			et.log.Debugf("Closing policy server client for %s:%d", portSpec.workload, portSpec.port)
+			portClient.cancel()
+		}
+		delete(et.policyWatches.watches, portSpec)
```
Comment (adleong) [inline controller/api/destination/endpoint_translator.go:504]: Similar to by comment above: what if there are multiple destination requests open which include the same pod+port?  If the pod is removed from one service's endpoints but remains in a second service's endpoints, I think this would close the policy watch.

File: controller/api/destination/server.go [hunk 17]
```diff
@@ -344,58 +324,132 @@ func (s *server) GetProfile(dest *pb.GetDestination, stream pb.Destination_GetPr
...
+
+	translator := newProfileTranslator(stream, s.log, "", port, endpoint)
+
+	// Create a client which watches for changes to the inbound policy for
+	// portSpec.
+	portSpec := &policyPb.PortSpec{
+		Workload: fmt.Sprintf("%s:%s", pod.Namespace, pod.Name),
+		Port:     port,
+	}
+	client, err := s.policyClient.WatchPort(stream.Context(), portSpec)
```
Comment (adleong) [inline controller/api/destination/server.go:370]: It's worth noting that this will create one WatchPort policy request for each GetProfile request.  This is probably okay for now, though for large clusters this could potentially create a very large number of redundant watches to the policy controller.

File: controller/api/destination/server.go [hunk 18]
```diff
@@ -344,58 +324,132 @@ func (s *server) GetProfile(dest *pb.GetDestination, stream pb.Destination_GetPr
...
+	// opaque, fall back to the opaque ports annotation on the pod.
+	//
+	// Shutdown is handled by destination server shutdown or the stream being
+	// closed.
+	for {
+		select {
+		case <-s.shutdown:
+			return nil
+		case <-stream.Context().Done():
+			// This case handles shutdown when updates will never be closed.
```
Comment (adleong) [inline controller/api/destination/server.go:415]: I don't really understand this comment.  It's hard to follow the scenario that it describes.  If the GetProfile request is cancelled, the policy request will also get cancelled because they share a context.  This should cause Recv to return an error and cause the updates stream to get closed.  We handle the channel being closed in the next case, so I don't think we actually need this case at all.

Comment (kleimkuhler) [reply controller/api/destination/server.go:415]: The explicit handling of stream closure here is needed because of the policy server `NotFound` check [here](https://github.com/linkerd/linkerd2/blob/52f791ad252cc843759afb68811e5cce14ad17ba/controller/api/destination/server.go#L386-L393).

In this case, the policy server updates goroutine has completed without closing `updates`. This means that this goroutine would never complete _if_ we are only checking for `updates` to be closed. So to prevent this, we have to explicitly handle `stream` closing which allow this goroutine to complete.

File: controller/api/destination/endpoint_translator.go (hunk 14 above)
Comment (kleimkuhler) [reply controller/api/destination/endpoint_translator.go:425]: Answered [here](https://github.com/linkerd/linkerd2/pull/6889#discussion_r720392607).

**Comment by adleong** [comment]
Ah, yeah, the fact that the `policyWatches` are scoped per translator make sense now.  So the `policyWatches` are really just to keep a hold of the streams so that they can be cancelled; there's no sharing of streams between different requests whatsoever.

This looks correct to me, but I still have misgivings about how it will behave on medium or large clusters.  Even a conservative case where 20 client replicas talk to 20 server replicas, this would cause the destination controller to create 400 streams to the policy controller.  Maybe that's fine, but I could also see this blowing up on non-trivial clusters.

Unfortunately, trying to share policy streams between different requests is quite complex, tricky, and annoying (as we have discussed).

Taking a step back, it may be possible to simplify the way this works by watching the Server resources via the Kubernetes API instead of using the policy controller API.  This allows us to leverage client-go's caching and watch management.

For instance, we could:

1. Add a proxyProtocol field to the endpoint watcher's `Address` struct
2. Have the endpoint watcher add an event handler for `Server` events
3. When a `Server` is updated, use the pod selector to find all affected pods and update their `proxyProtocol` field
4. When sending endpoint updates, take the `proxyProtocol` into consideration when computing the opaqueTransport

I know this is a very different approach to what's in this PR so I'm curious to hear what you think.  But I think an approach like this one would be a lot simpler because we don't need to manage starting and stopping streams dynamically and much more performant since we can utilize client-go to handle caching and sharing of watches.

**Comment by kleimkuhler** [comment]
Closing in favor of #7184 which uses the policy CRDs.
